### PR TITLE
Feature/2021 01 21 redis queue

### PR DESCRIPTION
**Pull request by mani-polarr**
Integrated Redis queue to SQS based event handler

### CODE REVIEW

File: lib/IntegrationEvent/Implementation/AWS-SNS-SQS/RedisHelper.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+            .setnx(activeEventKey, "1")
+            .expire(activeEventKey, 120)
+            .exec();
+          return eventDataString;
+        }
+      } else {
+        await options.RedisClient.multi()
+          .del(sortedSetKey, hashKey, activeSetKey)
+          .exec();
+        const queueIdList = await options.RedisClient.zpopmin(`master_event_queue`, 1);
```
Comment (VolatileBit) [inline lib/IntegrationEvent/Implementation/AWS-SNS-SQS/RedisHelper.ts]: @al-polarr Mani and I had a discussion about this. Basically, before events are published, the corresponding queue IDs are stored in Redis's master queue set. So there are two ways to deal with how to get the next queue to process when you finish processing all events of the current queue:

1. Mani's current method of handling this is to pop the earliest queue ID from the master queue, then process all events under that queue ID.
2. We can use SQS to initiate the processing of a queue. Considering that SQS is at least once delivery, the event itself can be seen as a mutex, once delivered, it should trigger the processing of a corresponding event queue.

I was thinking method 1 renders SQS effectively useless except in the cases where:

1. It is the first event arriving on a consumer, triggering the `fetchNextEvent` infinite loop.
2. Redis crashes, and we would rely on SQS's loosely FIFO event delivery, where if an event arrives and does not exist in Redis, the consumers will process it anyway.

My concern being method 1 might waste compute cycles by keep checking `fetchNextEvent`, and the master queue set in Redis becomes central orchestration point. If we scale up, we will end up with many consumers trying to `zpopmin` on the master queue set. Another point being technically SQS should be more reliable than Redis, since in Redis we are constrained by memory/sharding etc.

Instead I thought we should let events coming from SQS to initiate queue processing, when any event from a queue arrives, start processing all events in that queue, from the earliest to the oldest, and when the queue has emptied, exit the loop and wait for the next initiation. However, if there are any consistency issues, we may get events in the master queue that's never emptied. We may have to build a master queue police bot to check on master queue consistency, and if any events are not processed within a certain timeframe(say 10 minutes), process them and alert us.

We'd like to hear your thoughts on this, and are still brainstorming possible edge cases with either solutions.

Comment (al-polarr) [reply lib/IntegrationEvent/Implementation/AWS-SNS-SQS/RedisHelper.ts]: The need for coupling Redis and SQS is not immediately obvious from looking at this implementation.  It would help to have concrete examples of interaction patterns that give rise to these multiple queue IDs to provide understanding of how the events must subsequently be de-queued from multiple queue IDs to ensure correctness.  With that understanding in place, it becomes more straightforward to determine if the implementation is correct, and write automated tests to validate that correctness.  To your point about edge cases, enumerating interaction patterns in this way will also help us identify any gaps that might not be captured here.

Comment (VolatileBit) [reply lib/IntegrationEvent/Implementation/AWS-SNS-SQS/RedisHelper.ts]: @al-polarr Provided 2 examples in https://www.notion.so/polarr/Redis-FIFO-Queue-RFC-a4160df6fb9346cea9c8ee95396c5efd Let me know if this helps.

Comment (VolatileBit) [reply lib/IntegrationEvent/Implementation/AWS-SNS-SQS/RedisHelper.ts]: I've also scheduled a meeting for tomorrow so we can sync up about this.

File: lib/IntegrationEvent/Implementation/AWS-SNS-SQS/RedisHelper.ts [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+import { WrappedNodeRedisClient } from "handy-redis";
+
+export class RedisHelper {
+  public static async storeEventData(options: {
+    event: IntegrationEvent;
+    RedisClient: WrappedNodeRedisClient;
+  }): Promise<void> {
+    if (options.RedisClient && options.RedisClient.nodeRedis.connected && options.event && options.event.queueId) {
+      try {
+        const hashKey = `{${options.event.queueId}}_data`;
```
Comment (VolatileBit) [inline lib/IntegrationEvent/Implementation/AWS-SNS-SQS/RedisHelper.ts]: Even though this is not official, Redis documentations and tutorials often use colon `:` as a separator. And generally the category names are in front. Can we follow this convention instead?

So data:queueID and queue:queueID for example.

Keys like user:12345:device:abc is more intuitive to read than abc:device:12345:user.

Comment (mani-polarr) [reply lib/IntegrationEvent/Implementation/AWS-SNS-SQS/RedisHelper.ts]: Makes sense. I will update it.

Comment (mani-polarr) [reply lib/IntegrationEvent/Implementation/AWS-SNS-SQS/RedisHelper.ts]: This has been updated in [885425b](https://github.com/Polarrco/mediator/pull/1/commits/885425b47478d73a6e3efff779e6c43722cdee6e)

File: package.json [hunk 3]
```diff
@@ -1,5 +1,5 @@
 {
-  "name": "@polarrco/mediator",
+  "name": "@Polarrco/mediator",
   "version": "0.1.0",
```
Comment (VolatileBit) [inline package.json]: Let's increment the package version to 0.2.0

File: lib/IntegrationEvent/Implementation/AWS-SNS-SQS/RedisHelper.ts [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+            .setnx(activeEventKey, "1")
+            .expire(activeEventKey, 120)
+            .exec();
+          return eventDataString;
+        }
+      } else {
+        await options.RedisClient.multi()
+          .del(sortedSetKey, hashKey, activeSetKey)
+          .exec();
+        await options.RedisClient.zpopmin(`queue:master`, 1);
```
Comment (VolatileBit) [inline lib/IntegrationEvent/Implementation/AWS-SNS-SQS/RedisHelper.ts]: zpopmin will pop the earliest queue, but not necessarily the current queue that's finished right? Should we use `zrem(masterSetKey, sortedSetKey)` instead?

Comment (VolatileBit) [reply lib/IntegrationEvent/Implementation/AWS-SNS-SQS/RedisHelper.ts]: Also, why not chain `del` and `zpopmin` in the same multi() transaction?

Comment (mani-polarr) [reply lib/IntegrationEvent/Implementation/AWS-SNS-SQS/RedisHelper.ts]: Transaction won't work in cluster mode because master set could be in any of the shards. Transaction only works within a shard.

File: lib/IntegrationEvent/Implementation/AWS-SNS-SQS/SQSHelper.ts [hunk 5]
```diff
@@ -60,24 +68,7 @@ export class SQSHelper {
...
-          const [eventConstructor, eventHandlers] = subscription;
-          if (!eventConstructor || !eventHandlers || eventHandlers.length === 0) {
-            throw new Error(`There is no event handler respecting to this event: ${parsedBody.EventName}.`);
-          } else {
-            const event = new eventConstructor(parsedBody.EventBody);
-            for (const handler of eventHandlers) {
-              await handler.handle(event);
-            }
-          }
+          await SQSHelper.handleEvent(parsedBody, subscriptions, options.RedisClient);
```
Comment (VolatileBit) [inline lib/IntegrationEvent/Implementation/AWS-SNS-SQS/SQSHelper.ts]: Maybe add a try-catch around this? Otherwise if something goes wrong with event handling it would be an unhandled exception with little debugging details right?

Comment (mani-polarr) [reply lib/IntegrationEvent/Implementation/AWS-SNS-SQS/SQSHelper.ts]: We will only be able to console.log the error here, unless we add logger or slack integration. We have try-catch in all of our event handler classes in the backend. Wouldn't the error handling be more scenario-specific there?

Comment (VolatileBit) [reply lib/IntegrationEvent/Implementation/AWS-SNS-SQS/SQSHelper.ts]: try-catch in event handler classes is more of a good practice, it is possible we overlook it and forgot to try catch. Having it here acts as a global try-catch for all event handlers in case something errors got thrown out. console.log()errors will go to PM2 logs, which still helps us with debugging.

Comment (mani-polarr) [reply lib/IntegrationEvent/Implementation/AWS-SNS-SQS/SQSHelper.ts]: Added in latest commit

File: lib/IntegrationEvent/Implementation/AWS-SNS-SQS/RedisHelper.ts (hunk 4 above)
Comment (mani-polarr) [reply lib/IntegrationEvent/Implementation/AWS-SNS-SQS/RedisHelper.ts]: Switched zpopmin with zrem in latest commit.

**Review by VolatileBit — APPROVED** [review]
